src/neo4j_genai/filters.py

```diff
@@ -0,0 +1,349 @@
+
+
+class GteOperator(Operator):
+    CYPHER_OPERATOR = ">="
+
+
+class InOperator(Operator):
+    CYPHER_OPERATOR = "IN"
+
+    def cleaned_value(self, value):
+        for val in value:
+            if not isinstance(val, (str, int, float)):
+                raise ValueError(f"Unsupported type: {type(val)} for value: {val}")
+        return value
+
+
+class NinOperator(InOperator):
+    CYPHER_OPERATOR = "NOT IN"
+
+
+class LikeOperator(Operator):
+    CYPHER_OPERATOR = "CONTAINS"
+
+    def cleaned_value(self, value):
+        if not isinstance(value, str):
+            raise ValueError(f"Expected string value, got {type(value)}: {value}")
+        return value.rstrip("%")
+
+
+class ILikeOperator(LikeOperator):
+    def lhs(self, field):
+        return f"toLower({self.node_alias}.`{field}`)"
+
+    def cleaned_value(self, value):
+        value = super().cleaned_value(value)
+        return value.lower()
+
+
+OPERATOR_PREFIX = "$"
+
+OPERATOR_EQ = "$eq"
+OPERATOR_NE = "$ne"
+OPERATOR_LT = "$lt"
+OPERATOR_LTE = "$lte"
+OPERATOR_GT = "$gt"
+OPERATOR_GTE = "$gte"
+OPERATOR_BETWEEN = "$between"
+OPERATOR_IN = "$in"
+OPERATOR_NIN = "$nin"
+OPERATOR_LIKE = "$like"
+OPERATOR_ILIKE = "$ilike"
+
+OPERATOR_AND = "$and"
+OPERATOR_OR = "$or"
+
+COMPARISONS_TO_NATIVE = {
+    OPERATOR_EQ: EqOperator,
+    OPERATOR_NE: NeqOperator,
+    OPERATOR_LT: LtOperator,
+    OPERATOR_LTE: LteOperator,
+    OPERATOR_GT: GtOperator,
+    OPERATOR_GTE: GteOperator,
+    OPERATOR_IN: InOperator,
+    OPERATOR_NIN: NinOperator,
+    OPERATOR_LIKE: LikeOperator,
+    OPERATOR_ILIKE: ILikeOperator,
+}
+
+
+LOGICAL_OPERATORS = {OPERATOR_AND, OPERATOR_OR}
+
+SUPPORTED_OPERATORS = (
+    set(COMPARISONS_TO_NATIVE).union(LOGICAL_OPERATORS).union({OPERATOR_BETWEEN})
+)
+
+
+class ParameterStore:
+    """
+    Store parameters for a given query.
+    Determine the parameter name depending on a parameter counter
+    """
+
+    def __init__(self):
+        self._counter = Counter()
+        self.params = {}
+
+    def _get_params_name(self, key="param"):
+        """Find parameter name so that param names are unique.
+        This function adds a suffix to the key corresponding to the number
+        of times the key have been used in the query.
+        E.g.
+        node.age >= $param_0 AND node.age <= $param_1
+
+        Args:
+            key (str): The prefix for the parameter name
+        Returns:
+            The full unique parameter name
+        """
+        # key = slugify(key.replace(".", "_"), separator="_")
+        param_name = f"{key}_{self._counter[key]}"
+        self._counter[key] += 1
+        return param_name
+
+    def add(self, key, value):
+        """This function adds a new parameter to the param dict.
+        It returns the name of the parameter to be used as a placeholder
+        in the cypher query, e.g. $param_0"""
+        param_name = self._get_params_name()
+        self.params[param_name] = value
+        return param_name
```

Comment:
Should returning the `param_name` be in its own public method?

Reply:
TBH I don't see how, the parameter name is dependent on the order in which the parameters are added and since we have the increment each time the `add` method is called, how would you write a specific public method?